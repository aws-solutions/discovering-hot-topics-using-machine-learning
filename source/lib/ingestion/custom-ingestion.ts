#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from 'aws-cdk-lib';
import * as events from 'aws-cdk-lib/aws-events';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as kinesis from 'aws-cdk-lib/aws-kinesis';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import * as fs from 'fs';
import { S3ToEventBridgeToLambda } from '../s3-event-notification/s3-eventbridge-lambda';
import path = require('path');

export class CustomIngestion extends cdk.NestedStack {
    public readonly s3Bucket: s3.Bucket;

    constructor(scope: Construct, id: string, props: cdk.NestedStackProps) {
        super(scope, id, props);

        const _streamArn = new cdk.CfnParameter(this, 'StreamARN', {
            type: 'String',
            description:
                'The name of the stream where the records from custom ingestion should be published for analysis',
            allowedPattern: '^arn:\\S+:kinesis:\\S+:\\d{12}:stream/\\S+$',
            constraintDescription: 'Please provide the Amazon Kinesis Data Streams name'
        });

        const _accessLoggingBucket = new cdk.CfnParameter(this, 'S3AccessLoggingBucket', {
            type: 'String',
            description: 'The ARN of the access logging bucket that should be used for S3 upload bucket',
            allowedPattern: '^arn:\\S+:s3:\\S+:\\S+:*',
            constraintDescription: 'Please provide a valid ARN for the S3 access logging bucket'
        });

        const _integrationEventBus = new cdk.CfnParameter(this, 'IntegrationBus', {
            type: 'String',
            allowedPattern: '^arn:\\S+:events:\\S+:\\d{12}:event-bus/\\S+$',
            description:
                'The Event Bus to publish metadata information directly for storage and visualization. This metadata information' +
                ' does not process through the workflow engine (e.g. metadata about call analytics processing)'
        });

        const _metadataNS = new cdk.CfnParameter(this, 'MetadataNS', {
            type: 'String',
            constraintDescription: 'Please provide the namespace to used for metadata events published to event bus',
            description:
                'The namespace is used as part of the event pattern to route events to targets on the event bus'
        });

        const _stream = kinesis.Stream.fromStreamArn(this, 'PublishCommentsStream', _streamArn.valueAsString);
        const _loggingBucket = s3.Bucket.fromBucketArn(this, 'AccessLoggingBucket', _accessLoggingBucket.valueAsString);
        const _integrationBus = events.EventBus.fromEventBusArn(
            this,
            'AppIntegrationBus',
            _integrationEventBus.valueAsString
        );

        /*
        The following environment variables for lambda functiona is an example for excel based ingestion.
            let _lambdaEnv = {
                STREAM_NAME: _stream.streamName,
                INTEGRATION_BUS_NAME: _integrationBus.eventBusName,
                ID: "1",
                CREATED_DATE: "2",
                TEXT: "3",
                LANG: "4",
                ACCOUNT_NAME: "xlsx_file",
                PLATFORM: "custom"
            }
        Note the values of ID, CREATED_DATE, TEXT, and LANG are column numbers from which the value will be extracted. A sample excel file is also available
        at this path source/lambda/custom-ingestion/test/fixtures/mock_data_file.xlsx.
        */

        /*
        The following environment variables for lambda function are set up for Amazon Transcribe call analytics based ingestion of data.
        It can be used for any other JSON based ingestion as well. The source/lambda/custom-ingestion/file_processor module defines the available processor 
        types and how they can be invoked through source/lambda/custom-ingestion/file_processor/processor_factory.py
         */
        let _lambdaEnv = {
            STREAM_NAME: _stream.streamName,
            INTEGRATION_BUS_NAME: _integrationBus.eventBusName,
            PROCESSOR_TYPE: 'TRANSCRIBE_CALL_ANALYTICS',
            ID: 'GENERATE',
            CREATED_DATE: 'NOW',
            TEXT: 'Content',
            LANG: 'LanguageCode',
            ACCOUNT_NAME: 'call_analytics',
            SENTIMENT: 'Sentiment',
            NAMESPACE: _metadataNS.valueAsString,
            PLATFORM: 'customingestion',
            LIST_SELECTOR: 'Transcript'
        };

        const _s3ToEventBridgeToLambda = new S3ToEventBridgeToLambda(this, 'CustomIngestion', {
            lambdaFunctionProps: {
                runtime: lambda.Runtime.PYTHON_3_11,
                code: lambda.Code.fromAsset('lambda/ingestion-custom'),
                handler: 'lambda_function.handler',
                environment: _lambdaEnv,
                timeout: cdk.Duration.minutes(15),
                memorySize: 256
            },
            bucketProps: {
                versioned: false,
                serverAccessLogsBucket: _loggingBucket,
                serverAccessLogsPrefix: 'customingestion/'
            },
            s3LoggingBucket: _loggingBucket
        });
        this.s3Bucket = _s3ToEventBridgeToLambda.s3Bucket;

        _stream.grantWrite(_s3ToEventBridgeToLambda.lambdaFunction);
        _integrationBus.grantPutEventsTo(_s3ToEventBridgeToLambda.lambdaFunction);

        new iam.Policy(this, 'TagS3Object', {
            document: new iam.PolicyDocument({
                statements: [
                    new iam.PolicyStatement({
                        effect: iam.Effect.ALLOW,
                        actions: ['s3:PutObjectTagging'],
                        resources: [`${_s3ToEventBridgeToLambda.s3Bucket.bucketArn}/*`]
                    })
                ]
            })
        }).attachToRole(_s3ToEventBridgeToLambda.lambdaFunction.role!);

        const s3PolicyUpdateRole = new iam.Role(this, 'BucketPolicyCustomResource', {
            assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
            inlinePolicies: {
                LambdaFunctionServiceRolePolicy: new iam.PolicyDocument({
                    statements: [
                        new iam.PolicyStatement({
                            actions: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
                            resources: [
                                `arn:${cdk.Aws.PARTITION}:logs:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:log-group:/aws/lambda/*`
                            ]
                        }),
                        new iam.PolicyStatement({
                            actions: ['s3:PutBucketPolicy', 's3:GetBucketPolicy', 's3:PutBucketPublicAccessBlock'],
                            resources: [_loggingBucket.bucketArn]
                        })
                    ]
                })
            }
        });

        const s3PolicyUpdate = new lambda.Function(this, 'UpdateS3BucketPolicy', {
            handler: 'index.handler',
            runtime: lambda.Runtime.PYTHON_3_11,
            code: lambda.Code.fromInline(fs.readFileSync(path.join(__dirname, 'index.py'), 'utf8')),
            role: s3PolicyUpdateRole,
            tracing: lambda.Tracing.ACTIVE,
            timeout: cdk.Duration.minutes(15)
        });

        // prettier-ignore
        const customResourcePolicyUpdate = new cdk.CustomResource(this, 'UpdateS3Policy', {// NOSONAR - CDK construct initialization 
            resourceType: 'Custom::UpdateS3Policy',
            serviceToken: s3PolicyUpdate.functionArn,
            properties: {
                LOGGING_BUCKET_NAME: _loggingBucket.bucketName,
                SOURCE_BUCKET_NAME: this.s3Bucket.bucketName
            }
        });

        new cdk.CfnOutput(this, 'S3BucketToUploadData', {
            value: this.s3Bucket.bucketArn,
            description: 'Bucket location to upload source files for ingestion'
        });

        (s3PolicyUpdate.node.defaultChild as lambda.CfnFunction).addMetadata('cfn_nag', {
            rules_to_suppress: [
                {
                    'id': 'W89',
                    'reason': 'This is not a rule for the general case, just for specific use cases/industries'
                },
                {
                    'id': 'W92',
                    'reason': 'Impossible for us to define the correct concurrency for clients'
                }
            ]
        });

        (
            s3PolicyUpdateRole.node.tryFindChild('DefaultPolicy')!.node.tryFindChild('Resource') as iam.CfnPolicy
        ).addMetadata('cfn_nag', {
            rules_to_suppress: [
                {
                    'id': 'W12',
                    'reason':
                        'The cloudwatch policy narrows the permissions to the specific resources to the extent possible'
                }
            ]
        });
    }
}
