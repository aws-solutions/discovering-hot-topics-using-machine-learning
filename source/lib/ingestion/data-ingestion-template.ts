#!/usr/bin/env node
/**********************************************************************************************************************
* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
* SPDX-License-Identifier: Apache-2.0
 *********************************************************************************************************************/


import * as ddb from 'aws-cdk-lib/aws-dynamodb';
import * as events from 'aws-cdk-lib/aws-events';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { EventbridgeToLambda } from '@aws-solutions-constructs/aws-eventbridge-lambda';
import { LambdaToDynamoDB } from '@aws-solutions-constructs/aws-lambda-dynamodb';
import * as defaults from '@aws-solutions-constructs/core';

export interface LambdaAndDynamoDBProps {
    /**
     * Properties for the lambda function that triggers the event containing config information
     * for the resource target to process
     */
    readonly lambdaFunctionProps?: lambda.FunctionProps;
    /**
     * User provided instance of the lambda function that triggers the event containing config
     * information for the resource target to process. If this property is set, @functionProps
     * is ignored
     */
    readonly existingLambda?: lambda.Function
    /**
     * Properties for the dynamoDB table that contains config information for the lambda functions
     * to use. The table is not a requirement for creating the construct. If @tableProps and
     * @existingTable are not set, the table is not created
     */
    readonly tableProps?: ddb.TableProps;
    /**
     * User provided instance of instance of the DyanamoDB table, if this is set, the properties
     * provided in @tableProps is ignored. If @tableProps and @existingTable are not set, the table
     * is not created
     */
    readonly existingTable?: ddb.Table;
    /**
     * The key path in SSM from where the credentials are to be read for the lambda to execute the API
     */
    readonly credentialKeyPath?: string;
}

/**
 * Interface defintion for the trigger constructs for the messaging service. The primary
 * source of trigger is a lambda function to begin with but other AWS resources may be
 * considered to be added in the future
 */
export interface TriggerSourceProps extends LambdaAndDynamoDBProps {
}

/**
 * Interface definition for the targets to which event bus will route events based on defined rules
 */
export interface ResourceTargetProps extends LambdaAndDynamoDBProps {
}

/**
 * Properties for the Ingestion construct
 */
export interface DataIngestionTemplateProps {
    /**
     * Properties to define tirgger source
     */
    readonly source: TriggerSourceProps;
    /**
     * Properties to define the resource targets
     */
    readonly target: ResourceTargetProps;
    /**
     * User provided instance of event bus on AWS EventBridge. If this is set, @ingestionEventBusProps
     * is ignored
     */
    readonly existingIngestionEventBus?: events.EventBus;
    /**
     * Properties to create a custom event bus
     */
    readonly ingestionEventBusProps?: events.EventBusProps;
    /**
     * Properties to define Rules to route events to targets
     */
    readonly ingestionEventRuleProps: events.RuleProps;
}

/**
 * A data type created to allow multiple construct resources created within the function to be
 * returned to the caller
 */
interface LambdaDDBConstruct {
    lambdaFunction: lambda.Function,
    ddbTable?: ddb.Table
}

/**
 * Qualify/ categorize constructs to be part of source trigger or resource targets
 */
enum Qualifier {
    SOURCE = 'Source',
    TARGET = 'Target'
}

/**
 * This construct will create a custom event bus using @IngestionProps supplied. It also automatically creates a
 * DLQ and associates it with the event bus, if none exists.
 *
 * This construct also adds the role policy permissions required for the source trigger to publish messages
 * on the event bus and for the targets to be able to read from the event bus
 */
export class DataIngestionTemplate extends Construct {
    private _bus: events.EventBus;
    private _rule: events.Rule;
    private _lambda: { [qualifier: string]: lambda.Function } = {};
    private _dynamoDB: { [qualifier: string]: ddb.Table } = {};

    constructor(scope: Construct, id: string, props: DataIngestionTemplateProps) {
        super(scope, id);

        // start of building target/ query engine lambda, optional DDB, and add optional SSM policy (TARGET)
        const _eventTarget = this.buildCustomBusAndAddRules(props);
        this._lambda[Qualifier.TARGET] = _eventTarget.lambdaFunction;

        if (props.target.credentialKeyPath) {
            this._lambda[Qualifier.TARGET].addToRolePolicy(this.buildSSMPolicy(props.target.credentialKeyPath));
        }

        if (props.target.tableProps || props.target.existingTable) {
            const _target = this.buildLambdaAndDynamoDB({
                existingLambda: this._lambda[Qualifier.TARGET],
                tableProps: props.target.tableProps,
                existingTable: props.target.existingTable
            }, Qualifier.TARGET);

            this._lambda[Qualifier.TARGET] = _target.lambdaFunction
            if (_target.ddbTable) {
                this._dynamoDB[Qualifier.TARGET] = _target.ddbTable;
            }
        }
        // end of building target/ query engine lambda, optional DDB, and add optional SSM policy

        // start of building source trigger lambda and optional DDB (SOURCE)
        const _source = this.buildLambdaAndDynamoDB(props.source, Qualifier.SOURCE);
        _source.lambdaFunction.addEnvironment("EVENT_BUS_NAME", this._bus.eventBusName)
        this._lambda[Qualifier.SOURCE] = _source.lambdaFunction;
        if (_source.ddbTable) {
            this._dynamoDB[Qualifier.SOURCE] = _source.ddbTable;
        }
        this._lambda[Qualifier.SOURCE].addToRolePolicy(this.buildPutEventPolicy());
        // end of building source trigger and optional DDB
    }

    /**
     * This method builds a custom bus and adds Rules to route events to target resources
     *
     * @param props
     */
    private buildCustomBusAndAddRules(props: DataIngestionTemplateProps): EventbridgeToLambda {
        if (props.existingIngestionEventBus) {
            this._bus = props.existingIngestionEventBus;
        }
        else if (props.ingestionEventBusProps) {
            this._bus = new events.EventBus(this, 'IngestionBus', props.ingestionEventBusProps);
        } else {
            throw new Error('Either ingestionEventBusProps or existingIngestionEventBus has to be set. Both cannot be undefined');
        }

        const _target = new EventbridgeToLambda(this, 'RuleTargetLambda', {
            lambdaFunctionProps: props.target.lambdaFunctionProps,
            existingLambdaObj: props.target.existingLambda,
            eventRuleProps: {
                ...props.ingestionEventRuleProps,
                eventBus: this._bus,
                enabled: true
            }
        });

        this._rule = _target.eventsRule;
        return _target;
    }

    /**
     * This method builds the lambda function and optionally the DynamoDB table that stores state/ config
     * information based on the parameters passed
     *
     * @param props - properties to build lambda and DynamoDB
     * @param id - @Qualifier
     */
    private buildLambdaAndDynamoDB(props: LambdaAndDynamoDBProps, id: Qualifier): LambdaDDBConstruct {
        const _lambda = defaults.buildLambdaFunction(this, {
            existingLambdaObj: props.existingLambda,
            lambdaFunctionProps: props.lambdaFunctionProps
        });

        let _ddbTable: ddb.Table | undefined;
        if (props.tableProps || props.existingTable) {
            const lambdaToDDB = new LambdaToDynamoDB(this, id, {
                existingLambdaObj: _lambda,
                dynamoTableProps: props.tableProps,
                existingTableObj: props.existingTable,
                tableEnvironmentVariableName: `${id.toUpperCase()}_DDB_TABLE`
            });

            _ddbTable = lambdaToDDB.dynamoTable;
        }

        return {
            lambdaFunction: _lambda,
            ddbTable: _ddbTable
        };
    }

    /**
     * Create SSM policy to perform GetParameter on the key path provided
     *
     * @param keyPath
     */
    private buildSSMPolicy(keyPath: string): iam.PolicyStatement {
        return new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            resources: [`arn:${cdk.Aws.PARTITION}:ssm:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:parameter${keyPath}`],
            actions: ['ssm:GetParameter']
        });
    }

    private buildPutEventPolicy(): iam.PolicyStatement {
        return new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            resources: [this._bus.eventBusArn],
            actions: ['events:PutEvents']
        })
    }

    public get rule(): events.Rule {
        return this._rule;
    }

    public get bus(): events.EventBus {
        return this._bus;
    }

    public get sourceLambda(): lambda.Function {
        return this._lambda[Qualifier.SOURCE];
    }

    public get configTable(): ddb.Table {
        return this._dynamoDB[Qualifier.SOURCE];
    }

    public get targetLambda(): lambda.Function {
        return this._lambda[Qualifier.TARGET];
    }

    public get stateTable(): ddb.Table {
        return this._dynamoDB[Qualifier.TARGET];
    }
}
