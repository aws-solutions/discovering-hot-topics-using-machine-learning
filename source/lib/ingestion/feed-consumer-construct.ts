#!/usr/bin/env node
/**********************************************************************************************************************
* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
* SPDX-License-Identifier: Apache-2.0
 *********************************************************************************************************************/

import { Construct } from 'constructs';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as kinesis from 'aws-cdk-lib/aws-kinesis';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { SqsDlq } from 'aws-cdk-lib/aws-lambda-event-sources';
import * as cdk from 'aws-cdk-lib';
import { KinesisStreamsToLambda } from '@aws-solutions-constructs/aws-kinesisstreams-lambda';
import * as defaults from '@aws-solutions-constructs/core';

export interface FeedConsumerProps {
    readonly functionProps: lambda.FunctionProps;
    readonly batchSize?: number;
    readonly retryAttempts?: number;
    readonly shardCount?: number;
    readonly retentionPeriod?: cdk.Duration;
}

export class FeedConsumer extends Construct {

    private readonly feedConsumerStream: kinesis.IStream;
    private _consumerFunction: lambda.Function;

    constructor(scope: Construct, id: string, props: FeedConsumerProps) {
        super(scope, id);

        const sqsQueue = defaults.buildQueue(this, 'SqsDlqQueue', {}).queue;

        const ingestionStream = new KinesisStreamsToLambda(this, 'IngestionStream', {
            lambdaFunctionProps: props.functionProps,
            kinesisStreamProps: {
                encryption: kinesis.StreamEncryption.MANAGED,
                retentionPeriod: props.retentionPeriod,
                shardCount: props.shardCount
            },
            kinesisEventSourceProps: {
                startingPosition: lambda.StartingPosition.TRIM_HORIZON,
                batchSize: props.batchSize,
                retryAttempts: props.retryAttempts,
                onFailure: new SqsDlq(sqsQueue)
            }
        });

        this._consumerFunction = ingestionStream.lambdaFunction;
        this.feedConsumerStream = ingestionStream.kinesisStream;

        // This dependency mapping is required because the lambda role policy is getting created before the SQS DLQ. Since the DLQ
        // does not exist the policy creation failes in CFN.
        (this._consumerFunction.role?.node.tryFindChild('DefaultPolicy') as iam.Policy).node.addDependency(sqsQueue);
    }

    public get lambdaFunction(): lambda.Function {
        return this._consumerFunction;
    }

    public get kinesisStream(): kinesis.IStream {
        return this.feedConsumerStream;
    }
}
