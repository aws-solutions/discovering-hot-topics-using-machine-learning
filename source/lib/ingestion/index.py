import json
import logging
import os
import uuid

import boto3
import cfnresponse

RESOURCE_PROPERTIES = "ResourceProperties"
SOURCE_BUCKET_NAME = "SOURCE_BUCKET_NAME"
LOGGING_BUCKET_NAME = "LOGGING_BUCKET_NAME"
PHYSICAL_RESOURCE_ID = "PhysicalResourceId"
POLICY_SID = "customingestionAccessLogsPolicy"
STATEMENT = "Statement"

DEFAULT_LEVEL = "DEBUG"

s3 = boto3.client("s3")
logger = logging.getLogger(__name__)
logger.setLevel(os.environ.get("LOG_LEVEL", DEFAULT_LEVEL))


def handler(event, context):
    logger.debug(f"The event is {event}")
    physical_resource_id = None

    try:
        physical_resource_id = event.get(PHYSICAL_RESOURCE_ID, uuid.uuid4().hex[:8])

        if event["RequestType"] == "Create" or event["RequestType"] == "Update":
            source_bucket_name = event[RESOURCE_PROPERTIES][SOURCE_BUCKET_NAME]
            logging_bucket_name = event[RESOURCE_PROPERTIES][LOGGING_BUCKET_NAME]
            account_id = context.invoked_function_arn.split(":")[4]
            aws_partition = context.invoked_function_arn.split(":")[1]

            s3.put_public_access_block(
                Bucket=logging_bucket_name,
                PublicAccessBlockConfiguration={
                    "BlockPublicAcls": True,
                    "IgnorePublicAcls": True,
                    "BlockPublicPolicy": False,
                    "RestrictPublicBuckets": True,
                },
                ExpectedBucketOwner=account_id,
            )

            policy_string = s3.get_bucket_policy(Bucket=logging_bucket_name)["Policy"]
            bucket_policy = json.loads(policy_string)

            update_policy = True

            for statement in bucket_policy[STATEMENT]:
                logger.debug(f"Statement is {json.dumps(statement)}")
                if "Sid" in statement and statement["Sid"] == POLICY_SID:
                    if (
                        f"arn:{aws_partition}:s3:::{source_bucket_name}"
                        not in statement["Condition"]["ArnLike"]["aws:SourceArn"]
                    ):
                        logger.info("Bucket policy exists but with a different bucket, hence removing it")
                        bucket_policy[STATEMENT].remove(statement)
                        break
                    else:
                        logger.info("Bucket policy exists, hence update not required")
                        update_policy = False
                        break

            if update_policy:
                bucket_policy[STATEMENT].append(
                    {
                        "Action": "s3:PutObject",
                        "Condition": {
                            "ArnLike": {"aws:SourceArn": [f"arn:{aws_partition}:s3:::{source_bucket_name}"]},
                            "StringEquals": {"aws:SourceAccount": account_id},
                        },
                        "Effect": "Allow",
                        "Principal": {"Service": "logging.s3.amazonaws.com"},
                        "Resource": f"arn:{aws_partition}:s3:::{logging_bucket_name}/customingestion*",
                        "Sid": POLICY_SID,
                    }
                )

                s3.put_bucket_policy(Bucket=logging_bucket_name, Policy=json.dumps(bucket_policy))

            s3.put_public_access_block(
                Bucket=logging_bucket_name,
                PublicAccessBlockConfiguration={
                    "BlockPublicAcls": True,
                    "IgnorePublicAcls": True,
                    "BlockPublicPolicy": True,
                    "RestrictPublicBuckets": True,
                },
                ExpectedBucketOwner=account_id,
            )
            logger.info("Policy Updated successfully. Sending success response to CloudFormation")

        cfnresponse.send(event, context, cfnresponse.SUCCESS, {}, physical_resource_id)
    except Exception as ex:
        logger.error(f"Failed to update policy, error is {str(ex)}")
        cfnresponse.send(
            event, context, cfnresponse.FAILED, {}, physicalResourceId=physical_resource_id, reason=str(ex)
        )
