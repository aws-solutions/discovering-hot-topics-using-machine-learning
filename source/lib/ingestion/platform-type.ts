#!/usr/bin/env node
/**********************************************************************************************************************
* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
* SPDX-License-Identifier: Apache-2.0
 *********************************************************************************************************************/

export interface PlatformType {
    name: string,
    topicModelling: boolean
}