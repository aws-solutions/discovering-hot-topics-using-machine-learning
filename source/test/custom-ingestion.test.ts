/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Template, Match } from 'aws-cdk-lib/assertions';
import * as events from 'aws-cdk-lib/aws-events';
import * as kinesis from 'aws-cdk-lib/aws-kinesis';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as cdk from 'aws-cdk-lib';
import { CustomIngestion } from '../lib/ingestion/custom-ingestion';

test('Test custom ingestion nested stack', () => {
    const stack = new cdk.Stack();
    const _eventBus = new events.EventBus(stack, 'Bus');
    const s3AccessLoggingBucket = new s3.Bucket(stack, 'AccessLog', {
        versioned: false,
        encryption: s3.BucketEncryption.S3_MANAGED,
        removalPolicy: cdk.RemovalPolicy.RETAIN,
        blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
        enforceSSL: true
    });

    const _s3EventIntegration = new CustomIngestion(stack, 'testS3Event', {
        parameters: {
            'EventBus': _eventBus.eventBusArn,
            'StreamARN': new kinesis.Stream(stack, 'Stream', {}).streamArn
        }
    });
        Template.fromStack(stack).hasResource('AWS::CloudFormation::Stack',
        {
            Type: 'AWS::CloudFormation::Stack',
            Properties: {
                TemplateURL: {},
                Parameters: {
                    EventBus: {
                        'Fn::GetAtt': Match.anyValue()
                    },
                    StreamARN: {
                        'Fn::GetAtt': Match.anyValue()
                    }
                }
            },
            UpdateReplacePolicy: 'Delete',
            DeletionPolicy: 'Delete'
        });
});
