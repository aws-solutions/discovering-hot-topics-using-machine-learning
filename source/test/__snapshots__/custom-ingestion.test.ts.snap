// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test custom ingestion nested stack 1`] = `
{
  "Conditions": {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": {
      "Fn::Or": [
        {
          "Fn::Equals": [
            {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        {
          "Fn::Equals": [
            {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Parameters": {
    "AssetParameters1abc0fc8c8d9ecaa933fa5a28b2250fd168e9adc879df30b7edc099776a4a4b7ArtifactHashD52197FF": {
      "Description": "Artifact hash for asset "1abc0fc8c8d9ecaa933fa5a28b2250fd168e9adc879df30b7edc099776a4a4b7"",
      "Type": "String",
    },
    "AssetParameters1abc0fc8c8d9ecaa933fa5a28b2250fd168e9adc879df30b7edc099776a4a4b7S3Bucket9E565654": {
      "Description": "S3 bucket for asset "1abc0fc8c8d9ecaa933fa5a28b2250fd168e9adc879df30b7edc099776a4a4b7"",
      "Type": "String",
    },
    "AssetParameters1abc0fc8c8d9ecaa933fa5a28b2250fd168e9adc879df30b7edc099776a4a4b7S3VersionKey57F7B07B": {
      "Description": "S3 key for asset version "1abc0fc8c8d9ecaa933fa5a28b2250fd168e9adc879df30b7edc099776a4a4b7"",
      "Type": "String",
    },
    "AssetParameters52da8d6f3623417d24416267c50cbf9094db8cebbff1e3139e545620c38312beArtifactHash3B86D530": {
      "Description": "Artifact hash for asset "52da8d6f3623417d24416267c50cbf9094db8cebbff1e3139e545620c38312be"",
      "Type": "String",
    },
    "AssetParameters52da8d6f3623417d24416267c50cbf9094db8cebbff1e3139e545620c38312beS3Bucket3B125BBB": {
      "Description": "S3 bucket for asset "52da8d6f3623417d24416267c50cbf9094db8cebbff1e3139e545620c38312be"",
      "Type": "String",
    },
    "AssetParameters52da8d6f3623417d24416267c50cbf9094db8cebbff1e3139e545620c38312beS3VersionKey64BA2168": {
      "Description": "S3 key for asset version "52da8d6f3623417d24416267c50cbf9094db8cebbff1e3139e545620c38312be"",
      "Type": "String",
    },
  },
  "Resources": {
    "BusEA82B648": {
      "Properties": {
        "Name": "Bus",
      },
      "Type": "AWS::Events::EventBus",
    },
    "Stream790BDEE4": {
      "Properties": {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": {
          "Fn::If": [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            {
              "Ref": "AWS::NoValue",
            },
            {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
        "StreamModeDetails": {
          "StreamMode": "PROVISIONED",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "testS3EventNestedStacktestS3EventNestedStackResource0A9F329B": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Parameters": {
          "EventBus": {
            "Fn::GetAtt": [
              "BusEA82B648",
              "Arn",
            ],
          },
          "StreamARN": {
            "Fn::GetAtt": [
              "Stream790BDEE4",
              "Arn",
            ],
          },
          "referencetoAssetParameters1abc0fc8c8d9ecaa933fa5a28b2250fd168e9adc879df30b7edc099776a4a4b7S3Bucket9E565654Ref": {
            "Ref": "AssetParameters1abc0fc8c8d9ecaa933fa5a28b2250fd168e9adc879df30b7edc099776a4a4b7S3Bucket9E565654",
          },
          "referencetoAssetParameters1abc0fc8c8d9ecaa933fa5a28b2250fd168e9adc879df30b7edc099776a4a4b7S3VersionKey57F7B07BRef": {
            "Ref": "AssetParameters1abc0fc8c8d9ecaa933fa5a28b2250fd168e9adc879df30b7edc099776a4a4b7S3VersionKey57F7B07B",
          },
        },
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://s3.",
              {
                "Ref": "AWS::Region",
              },
              ".",
              {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              {
                "Ref": "AssetParameters52da8d6f3623417d24416267c50cbf9094db8cebbff1e3139e545620c38312beS3Bucket3B125BBB",
              },
              "/",
              {
                "Fn::Select": [
                  0,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "AssetParameters52da8d6f3623417d24416267c50cbf9094db8cebbff1e3139e545620c38312beS3VersionKey64BA2168",
                      },
                    ],
                  },
                ],
              },
              {
                "Fn::Select": [
                  1,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "AssetParameters52da8d6f3623417d24416267c50cbf9094db8cebbff1e3139e545620c38312beS3VersionKey64BA2168",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
